basic funcionality + ssl

jsvd opened the pull request:
this pr implements logstash-input-http using Puma server.
Has basic functionality of handling HTTP requests and also supports SSL through java keystores.

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+  config :port, :validate => :number, :default => 8080
+
+  # Maximum number of threads to use
+  config :max_threads, :validate => :number, :default => 4
```
Even if this class doesn't inherits from `LogStash::Inputs::Threadable` we should use the same option `threads`

ph replied:
We are also using the `workers` concept for the filter/outputs.

jsvd replied:
but since this is a Puma-specific setting, maybe I should change this to a more puma-specific name like puma_max_threads?

ph replied:
Even if it is puma specific, for a user more threads (or more workers) mean more throughput. They don't have to worry about the implementation details this is why I don't think we should use `puma_*`.

ph replied:
Also for the number of threads, what is a good number? Should we make that assumption with the number of CPUs?

Concurrent-ruby provides a cross platform way of getting that information. https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/utility/processor_count.rb

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+      new_hash[k.downcase] = v
+    end
+    new_hash
+  end
```
You can use `#each_with_object` If you need to save your data in a new hash.

```ruby
m = { "awe" => 1, "weeAAAAee" => 2 }
m.each_with_object({}) { |(k,v), memo| memo[k.downcase] = v }
#=> {"awe"=>1, "weeaaaaee"=>2}

jsvd replied:
I usually avoid inject or each with object as I find them hard to read and reason about. strong feelings on this?

ph replied:
No strong feeling on this, Since they are party of ruby I've been using them.

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+  def run(queue)
+    @server.app = Proc.new do |req|
+      begin
+        REJECTED_HEADERS.each {|k| req.delete(k) }
```
I am wondering if it would be better to takes the headers from the req and to check if the `REJECTED_HEADERS` include it. I assume we have usually less keys in the request than the rejected list, I don't know rack enough if it is the case.

```ruby
m.delete_if { |k,v| REJECTED_HEADERS.include?(k) }
```

jsvd replied:
well. complexity wise it's the same, you have to do MxN checks

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+      @server.add_ssl_listener @host, @port, ctx
+    else
+      @server.add_tcp_listener @host, @port
+    end
```
Missing the ()
```ruby
@server.add_tcp_listener(@host, @port)
```

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+  # mostly due to rack compliance
+  REJECTED_HEADERS = ["puma.socket", "rack.hijack?", "rack.hijack", "rack.url_scheme", "rack.after_reply", "rack.version", "rack.errors", "rack.multithread", "rack.multiprocess", "rack.run_once", "SCRIPT_NAME", "QUERY_STRING", "SERVER_PROTOCOL", "SERVER_SOFTWARE", "GATEWAY_INTERFACE"]
+
+  RESPONSE_HEADERS = {'Content-Type' => 'text/html'}
```
Since we aren't returning html, we should use `text/plain`

jsvd replied:
:+1:

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+      rescue => e
+        @logger.error("unable to process event #{req.inspect}. exception => #{e.inspect}")
+        ['500', RESPONSE_HEADERS, ['internal error']]
+      end
```
Thinking out loud here.

Since our http input is in reality a small webservice, we should consider the response format.

If we return a response to the callback we should use the same format as the request.
If they post a json document we should return a json document, same situation with xml.

To make things less complicated.. I dont think we need to return a string, the http code are enough in this case and more flexible.

jsvd replied:
yep this is something I want do take care of next: properly use the request content-type to choose the codec and adapt the response.

ph replied:
ok created a ticket for this #3

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+      ctx.keystore = @keystore
+      ctx.keystore_pass = @keystore_password.value
+      @server.add_ssl_listener @host, @port, ctx
+    else
```
in the SSL context, `@keystore*` becomes mandatory, we should add some sort of validation.

ph commented on lib/logstash/inputs/http.rb:
```diff
@@ -0,0 +1,100 @@
+  #TODO: config :cacert, :validate => :path
+
+  # Set the truststore password
+  config :keystore_password, :validate => :password
```
Concerning keystore/keystore_password can we use the same option name as the lumberjack input?

jsvd replied:
it will eventually support simple .crt files as well so keystores are a separate config parameter

ph replied:
Maybe point to some doc to setup ssl? I've personally never used keystore jks before.

ph commented:
Minor comment on the SSL, LGTM, tested the normal http not the ssl version.

jsvd commented:
Addressed most of the issues, @ph 
Thanks to the issues you raised, the specs I wrote showed the teardown sequence wasn't properly closing the tcp socket. it's now fixed!

ph commented:
wohoo :)
LGTM! :shipit:

For a future enhancement:

@jsvd I was tinkering with the input this weekend, I think we should add a way to support tagging of incoming web calls. I see something like this on the caller side.

```
curl 127.0.0.1:8080/?tags=github,logstash
#OR
curl 127.0.0.1:8080/tags/github,logstash
```

And we add the tags to the generated events so we use 1 input for multiple incoming webhooks, I think this could be a future enhancement.

jsvd commented:
@ph yep tagging is essential for conditionals. One very nice way of tagging would be to use the headers on the caller side. Since headers will be exploded to the "headers" field they can then be used in conditionals automatically.
Alternatively, we can have a config setting that, when turned on, will parse the query_parameters and construct an hash out of them.

ph commented:
@jsvd headers work if the user control the caller part, yes we need them +1 :)
@logstash-plugins/owners I think we should include this plugin as part of the defaults plugins.
